### PR TITLE
Define isnan for Number

### PR DESCRIPTION
**Pull request by bkamins**
Currently `isnan` is defined for `Real` and for `Complex` which are only subtypes of `Number` in Base.

However, in general it is possible that there are numbers that are subtype of `Number` that are not `Real` or `Complex` so it would be good to have `isnan` defined by default for all `Number` to avoid having writing code like:
```
function fun(x::Number)
    if x isa Union{Real, Complex)
        isnan(x)
    else
        # do something special as it is not guaranteed that isnan is defined here
    end
end
```
(or the equivalent of the above using multiple dispatch)

### CODE REVIEW

**Review by simonbyrne — APPROVED** [review]

File: base/float.jl [hunk 1]
```diff
@@ -447,7 +447,7 @@ Test whether a number value is a NaN, an indeterminate value which is neither an
 nor a finite number ("not a number").
 """
 isnan(x::AbstractFloat) = (x != x)::Bool
-isnan(x::Real) = false
+isnan(x::Number) = false
 
 isfinite(x::AbstractFloat) = x - x == 0
```
Comment (jw3126) [inline base/float.jl:452]: How about `isfinite`? I think it is also not defined for `Number`? 

Also a test would be good so this does not get deleted by accident. 
```julia
struct N38232 <: Number end
@test isfinite(N38232())
@test !isnan(N38232())
```

Comment (bkamins) [reply base/float.jl:452]: `isfinite` is not `true` by default. It is using `Base.decompose` for `Real` which is not exposed therefore I would leave the decision what to do it for a separate PR (fell free to open it; I need `isnan` for DataFrames.jl, as `NaN` has a special treatment in comparisons).

I will add the test you suggest - thank you.

**Comment by StefanKarpinski** [comment]
I'm a bit late to the party here, but I think it would be fine to just broaden this definition:
```jl
isnan(x::Number) = (x != x)::Bool
```
Non-reflexiveness is pretty much the defining characteristic of NaNs. There's some concern about performance, but I think that it should mostly be ok and specialized methods can always be added for performance as needed.

**Comment by bkamins** [comment]
Makes sense. The only issue is that it would be breaking in corner cases, as the old definition for `Real` was different (probably this can be treated as a bug-fix though).

**Pull request merged**